File: adv20/23.py

def extract3(cups, current):
    n = current
    e = []
    for _ in range(3):
        n = cups[n]
        e.append(n)
    cups[current] = cups[n]
    return e
def insert3(cups, dest, picked):
    n = cups[dest]
    cups[dest] = picked[0]
    cups[picked[2]] = n

def play(cups, rounds, maxC, start):

    current = start
    picked = None

    for _ in range(rounds):
        picked = extract3(cups, current)
        dest = current-1 if current > 1 else maxC
        while dest in picked:
            dest = dest-1 if dest > 1 else maxC
        insert3(cups, dest, picked)
        current = cups[current]
    return cups

data = [int(x) for x in list("364289715")]
cups = [x+1 for x in range(len(data)+1)]
for i in range(len(data)-1):
    cups[data[i]] = data[i+1]
cups[data[-1]] = data[0]

cups = play(cups, 10, 9, data[0])
i = 1
while cups[i] != 1:
    print(cups[i], end='')
    i = cups[i]
print()

cups = [x+1 for x in range(1000001)]
for i in range(len(data)-1):
    cups[data[i]] = data[i+1]
cups[-1] = data[0]
cups[data[-1]] = 10

cups = play(cups, 10000000, 1000000, data[0])
print(cups[1]*cups[cups[1]])

File: adv20/23.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void extract3(int* cups, int* extracted, int current)
{
  int n = current;
  for(int i = 0; i < 3; ++i) {
    n = cups[n];
    extracted[i] = n;
  }
  cups[current] = cups[n];
}

void insert3(int* cups, int dest, int* picked)
{
  int n = cups[dest];
  cups[dest] = picked[0];
  cups[picked[2]] = n;
}

void play(int* cups, int rounds, int maxC, int start)
{
  int current = start;
  int picked[3];
  int dest = 0;

  for(int i = 0; i < rounds; ++i)
  {
    extract3(cups, picked, current);
    if(current > 1)
      dest = current-1;
    else
      dest = maxC;
    while(dest == picked[0] || dest == picked[1] || dest == picked[2])
      if(--dest == 0)
        dest = maxC;
    insert3(cups, dest, picked);
    current = cups[current];

  }
}

int main()
{
  char* data = "364289715";
  int d_len = strlen(data);

  int* cups = malloc(sizeof(int)*1000001);
  for(int i = 0; i < 1000000; ++i)
    cups[i] = i+1;
  for(int i = 0; i < d_len-1; ++i)
    cups[data[i]-48] = data[i+1]-48;
  cups[1000000] = data[0]-48;
  cups[data[d_len-1]-48] = 10;

  play(cups, 10000000, 1000000, data[0]-48);
  printf("%ld\n", (long)cups[1]*cups[cups[1]]);
}
